src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java

```diff
@@ -129,19 +138,62 @@ public void run(ActionContext context) throws Exception {
+  }
 
-    LOG.info("Executing SQL as job {}.", jobId.getJob());
-    LOG.debug("The BigQuery SQL is {}", config.getSql());
+  protected void executeQueryWithExponentialBackoff(BigQuery bigQuery,
+                                                   QueryJobConfiguration queryConfig, ActionContext context) {
+    Failsafe.with(getRetryPolicy()).run(() -> executeQuery(bigQuery, queryConfig, context, retryOnReason));
+  }
 
-    // Wait for the query to complete
-    queryJob = queryJob.waitFor();
+  private RetryPolicy<Object> getRetryPolicy() {
+    // Exponential backoff with initial retry of 1 second and max retry of 32 seconds.
+    Duration initialRetryDuration = Duration.ofSeconds(1);
+    Duration maxRetryDuration = Duration.ofSeconds(32);
+    int multiplier = 2;
+    int maxRetryCount = 5;
+    return RetryPolicy.builder()
+            .handle(BigQueryJobExecutionException.class)
+            .withBackoff(initialRetryDuration, maxRetryDuration, multiplier)
+            .withMaxRetries(maxRetryCount)
+            .onRetry(event -> LOG.debug("Retrying BigQuery Execute job. Retry count: {}", event.getAttemptCount()))
+            .onSuccess(event -> LOG.debug("BigQuery Execute job executed successfully."))
+            .onRetriesExceeded(event -> LOG.error("Retry limit reached for BigQuery Execute job."))
+            .build();
+  }
+
+  private void executeQuery(BigQuery bigQuery, QueryJobConfiguration queryConfig, ActionContext context,
+                            Set<String> retryOnReason) throws InterruptedException, BigQueryJobExecutionException {
+    // Location must match that of the dataset(s) referenced in the query.
+    JobId jobId = JobId.newBuilder().setRandomJob().setLocation(config.getLocation()).build();
+    Job queryJob;
+
+    try {
+      queryJob = bigQuery.create(JobInfo.newBuilder(queryConfig).setJobId(jobId).build());
+
+      LOG.info("Executing SQL as job {}.", jobId.getJob());
+      LOG.debug("The BigQuery SQL is {}", config.getSql());
+
+      // Wait for the query to complete
+      queryJob = queryJob.waitFor();
+    } catch (BigQueryException e) {
+      LOG.error("The query job {} failed. Error: {}", jobId.getJob(), e.getMessage());
+      if (retryOnReason.contains(e.getError().getReason())) {
+        throw new BigQueryJobExecutionException(e.getMessage());
+      }
+      throw new RuntimeException(e);
+    }
 
     // Check for errors
     if (queryJob.getStatus().getError() != null) {
       // You can also look at queryJob.getStatus().getExecutionErrors() for all
       // errors, not just the latest one.
-      throw new RuntimeException(queryJob.getStatus().getExecutionErrors().toString());
+      LOG.error("The query job {} failed. Error: {}", jobId.getJob(), queryJob.getStatus().getError());
+      if (retryOnReason.contains(queryJob.getStatus().getError().getReason())) {
+        throw new BigQueryJobExecutionException(queryJob.getStatus().getError().getMessage());
```

Comment:
It's a checked exception. What would it be converted to when we are out of retries?